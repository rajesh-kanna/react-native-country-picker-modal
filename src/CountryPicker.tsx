import React, { ReactNode, useState, useEffect } from 'react'
import { ModalProps, FlatListProps } from 'react-native'
import { CountryModal } from './CountryModal'
import { HeaderModal } from './HeaderModal'
import { Country, CountryCode, FlagType } from './types'
import { CountryFilter, CountryFilterProps } from './CountryFilter'
import { FlagButton } from './FlagButton'
import { useContext } from './CountryContext'
import { CountryList } from './CountryList'

interface State {
  visible: boolean
  countries: Country[]
  filter?: string
}

const renderFlagButton = (
  props: FlagButton['props'] & CountryPickerProps['renderFlagButton']
): ReactNode =>
  props.renderFlagButton ? (
    props.renderFlagButton(props)
  ) : (
    <FlagButton {...props} />
  )

const renderFilter = (
  props: CountryFilter['props'] & CountryPickerProps['renderCountryFilter']
): ReactNode =>
  props.renderCountryFilter ? (
    props.renderCountryFilter(props)
  ) : (
    <CountryFilter {...props} />
  )

interface CountryPickerProps {
  countryCode: CountryCode
  modalProps?: ModalProps
  filterProps?: CountryFilterProps
  flatListProps?: FlatListProps<Country>
  withEmoji?: boolean
  withCountryName?: boolean
  withFilter?: boolean
  withAlphaFilter?: boolean
  withCallingCode?: boolean
  withCurrency?: boolean
  withFlag?: boolean
  withModal?: boolean
  visible?: boolean
  renderFlagButton?(props: FlagButton['props']): ReactNode
  renderCountryFilter?(props: CountryFilter['props']): ReactNode
  onSelect(country: Country): void
}

export const CountryPicker = (props: CountryPickerProps) => {
  const {
    countryCode,
    renderFlagButton: renderButton,
    renderCountryFilter,
    filterProps,
    modalProps,
    flatListProps,
    onSelect,
    withEmoji,
    withFilter,
    withCountryName,
    withAlphaFilter,
    withCallingCode,
    withCurrency,
    withFlag,
    withModal
  } = props
  const [state, setState] = useState<State>({
    visible: props.visible || false,
    countries: [],
    filter: ''
  })
  const { translation, getCountries } = useContext()
  const { visible, filter, countries } = state
  const onOpen = () => setState({ ...state, visible: true })
  const onClose = () => setState({ ...state, filter: '', visible: false })
  const setFilter = (filter: string) => setState({ ...state, filter })
  const setCountries = (countries: Country[]) =>
    setState({ ...state, countries })
  const onSelectClose = (country: Country) => {
    onSelect(country)
    onClose()
  }
  const flagProp = {
    withEmoji,
    withCountryName,
    countryCode,
    renderFlagButton: renderButton,
    onOpen
  }
  useEffect(() => {
    const countries = getCountries(
      withEmoji ? FlagType.EMOJI : FlagType.FLAT,
      translation
    )
    setCountries(countries)
  }, [translation, withEmoji])

  return (
    <>
      {renderFlagButton(flagProp)}
      <CountryModal
        {...{ ...modalProps, visible, withModal }}
        onRequestClose={onClose}
      >
        <HeaderModal
          {...{ withFilter, onClose }}
          renderFilter={(props: CountryFilter['props']) =>
            renderFilter({
              ...props,
              renderCountryFilter,
              onChangeText: setFilter,
              value: filter,
              ...filterProps
            })
          }
        />
        <CountryList
          {...{
            onSelect: onSelectClose,
            data: countries,
            letters: [],
            withAlphaFilter: withAlphaFilter && filter === '',
            withCallingCode,
            withCurrency,
            withFlag,
            withEmoji,
            filter,
            flatListProps
          }}
        />
      </CountryModal>
    </>
  )
}

CountryPicker.defaultProps = {
  withAlphaFilter: false,
  withCallingCode: false
}
