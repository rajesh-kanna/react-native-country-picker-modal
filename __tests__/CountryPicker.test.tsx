import React from 'react'
// import 'react-native'

import renderer from 'react-test-renderer'

import CountryPicker from '../src/'

it('CountryPicker can be created', () => {
  const picker = renderer.create(
    <CountryPicker countryCode={'US'} onSelect={() => {}} />
  )
  expect(picker).toBeDefined()
})

it('<CountryPicker /> toMatchSnapshot', () => {
  const tree = renderer
    .create(<CountryPicker countryCode={'US'} onSelect={() => {}} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
