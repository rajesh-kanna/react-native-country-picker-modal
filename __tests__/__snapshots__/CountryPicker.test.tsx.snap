// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CountryPicker /> toMatchSnapshot 1`] = `
Array [
  <View
    accessible={true}
    isTVSelectable={true}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "flex-start",
            "justifyContent": "flex-start",
          },
          Object {
            "marginTop": 0,
          },
          undefined,
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "marginRight": 10,
            "width": 30,
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "transparent",
              "borderWidth": 0.5,
              "fontSize": 30,
              "justifyContent": "center",
            }
          }
        >
          <Text
            allowFontScaling={false}
          >
            ðŸ‡ºðŸ‡¸
          </Text>
        </Text>
      </View>
    </View>
  </View>,
  <Modal
    animated={true}
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    visible={false}
  >
    <RCTSafeAreaView
      emulateUnlessSupported={true}
      style={
        Array [
          Object {
            "flex": 1,
          },
          Object {
            "backgroundColor": "#ffffff",
          },
        ]
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "height": 48,
                "justifyContent": "center",
                "width": "15%",
              },
              undefined,
            ]
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../../src/assets/images/close.ios.png",
                }
              }
              style={
                Array [
                  Object {
                    "height": 25,
                    "resizeMode": "contain",
                    "width": 25,
                  },
                  undefined,
                  Object {
                    "tintColor": "#000000",
                  },
                ]
              }
            />
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "flex": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
            },
            Object {
              "backgroundColor": "#ffffff",
            },
          ]
        }
      >
        <RCTScrollView
          ItemSeparatorComponent={[Function]}
          automaticallyAdjustContentInsets={false}
          data={Array []}
          disableVirtualization={false}
          getItem={[Function]}
          getItemCount={[Function]}
          getItemLayout={[Function]}
          horizontal={false}
          initialNumToRender={12}
          keyExtractor={[Function]}
          keyboardShouldPersistTaps="handled"
          maxToRenderPerBatch={10}
          numColumns={1}
          onContentSizeChange={[Function]}
          onEndReachedThreshold={2}
          onLayout={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          onScrollToIndexFailed={[Function]}
          removeClippedSubviews={false}
          renderItem={[Function]}
          scrollEventThrottle={1}
          stickyHeaderIndices={Array []}
          testID="list-countries"
          updateCellsBatchingPeriod={50}
          viewabilityConfigCallbackPairs={Array []}
          windowSize={21}
        >
          <View />
        </RCTScrollView>
      </View>
    </RCTSafeAreaView>
  </Modal>,
]
`;
